patches/server/0933-Add-Offline-PDC-API.patch

```diff
@@ -0,0 +1,212 @@
+             }
++            if (!(this instanceof Player)) { // Paper - normal logic for non player
+             this.getBukkitEntity().readBukkitValues(nbt);
++            } // Paper
+             if (nbt.contains("Bukkit.invisible")) {
+                 boolean bukkitInvisible = nbt.getBoolean("Bukkit.invisible");
+                 this.setInvisible(bukkitInvisible);
+diff --git a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
+index 601f8099f74e81c17600566b3c9b7a6dd39c9bcb..a68c11358e4a87a9ee8814e4044414a453343a8b 100644
+--- a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
++++ b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
+@@ -32,21 +32,37 @@ public class PlayerDataStorage {
+     }
+ 
+     public void save(Player player) {
++        // Paper start
++        CompoundTag nbttagcompound = player.saveWithoutId(new CompoundTag());
++        save(player.getUUID(), nbttagcompound, player);
++    }
++    public void save(java.util.UUID uuid, CompoundTag tag, @Nullable Player player) {
+         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
+         try {
+-            CompoundTag nbttagcompound = player.saveWithoutId(new CompoundTag());
+-            File file = File.createTempFile(player.getStringUUID() + "-", ".dat", this.playerDir);
++            File file = File.createTempFile(uuid + "-", ".dat", this.playerDir);
+ 
+-            NbtIo.writeCompressed(nbttagcompound, file);
+-            File file1 = new File(this.playerDir, player.getStringUUID() + ".dat");
+-            File file2 = new File(this.playerDir, player.getStringUUID() + ".dat_old");
++            NbtIo.writeCompressed(tag, file);
++            File file1 = new File(this.playerDir, uuid + ".dat");
++            File file2 = new File(this.playerDir, uuid + ".dat_old");
+ 
+             Util.safeReplaceFile(file1, file, file2);
+         } catch (Exception exception) {
+-            PlayerDataStorage.LOGGER.warn("Failed to save player data for {}", player.getScoreboardName(), exception); // Paper
++            if (player == null) {
++                PlayerDataStorage.LOGGER.warn("Failed to save offline player data for {}", uuid, exception); // Paper
++            } else {
++                PlayerDataStorage.LOGGER.warn("Failed to save player data for {}", player.getScoreboardName(), exception); // Paper
++            }
+         }
+-
+     }
++    // check if an offline PDC has already been created before its first join
++    public static boolean checkOfflinePDC(Player player, CompoundTag nbt, boolean checkOnly) {
++        boolean result = checkOnly ? player.getBukkitEntity().searchAvailableData(nbt) : player.getBukkitEntity().readBukkitValues(nbt);
++        if (result && nbt.size() == 1) {
++            return true;
++        }
++        return false;
++    }
++    // Paper end
```

Comment:
Be more strict with your `// Paper start` and `// Paper end` comments. Most of the time, they should only be surrounding actual diff, not a big chunk with several line changes inside it.

Reply:
Resolved, removed some useless diff